### PR TITLE
[Bug] Render transform logic on rule build

### LINKED ISSUE #2740: [Bug] OSQuery transformation not included in release packages
**Describe the bug**
Rules are capable of including a `transform` field in them which includes OSQueries or Insight queries. These are then transformed and replace anchors in the investigation guides.

It appears that some of these did not transform properly in released packages, which requires a bug fix and patch release. 

Screenshot 2023-04-24 at 12 15 21 PM

### PR DESCRIPTION
**Pull request by brokensound77**
## Issues
resolves #2740 

## Summary
This bug actually led to the discovery of 2 additional bugs

- A: `TOMLRuleContents.to_dict` was not called from `TOMLRuleContents.to_api_format`, bypassing the transform logic (original bug)
- B: `test_transform_fields` tests were calling the `TOMLRuleContents.to_dict` rather than `TOMLRuleContents.to_api_format` method, bypassing the emulated check on actual rule generation
- C: for the substitution rendering, there was the potential for pattern poisoning where any word within the investigation guide starting with `$` would render as a pattern and raise a key error for missing values. 

## Details
- A: resolved by making the call
- B: resolved by fixing the test
- C: resolved by using `safe_substitute` to ignore non-explicit patterns

## Testing

- `make test` and view `test_transform_fields`
- `make release` and search for `$osquery_` pattern in `extras/rules` as well as `extras/fleet/kibana`

<details>

image

</details>

### CODE REVIEW

File: detection_rules/rule.py [hunk 1]
```diff
@@ -352,7 +352,7 @@ def process_note_plugins():
                 rendered_patterns.update(**{f'{plugin}_{i}': e for i, e in enumerate(entries)})
 
             note_template = PatchedTemplate(note)
-            rendered_note = note_template.substitute(**rendered_patterns)
+            rendered_note = note_template.safe_substitute(**rendered_patterns)
```
Comment (brokensound77) [inline detection_rules/rule.py:355]: fix C

File: detection_rules/rule.py [hunk 2]
```diff
@@ -1107,8 +1107,8 @@ def flattened_dict(self) -> dict:
 
     def to_api_format(self, include_version=True) -> dict:
         """Convert the TOML rule to the API format."""
-        converted = self.data.to_dict()
-        converted = self._post_dict_conversion(converted)
+        converted_data = self.to_dict()['rule']
```
Comment (brokensound77) [inline detection_rules/rule.py:1110]: fix A

File: tests/test_transform_fields.py [hunk 3]
```diff
@@ -92,7 +92,7 @@ def test_transform_guide_markdown_plugins(self) -> None:
 
         new_rule_contents = TOMLRuleContents.from_dict(rule_dict)
         new_rule = TOMLRule(path=sample_rule.path, contents=new_rule_contents)
-        rendered_note = new_rule.contents.to_dict()['rule']['note']
+        rendered_note = new_rule.contents.to_api_format()['note']
```
Comment (brokensound77) [inline tests/test_transform_fields.py:95]: fix B

File: tests/test_transform_fields.py [hunk 4]
```diff
@@ -109,6 +109,6 @@ def test_plugin_conversion(self):
             rule_dict_copy.update(**transform)
             new_rule_contents = TOMLRuleContents.from_dict(rule_dict_copy)
             new_rule = TOMLRule(path=sample_rule.path, contents=new_rule_contents)
-            rendered_note = new_rule.contents.to_dict()['rule']['note']
+            rendered_note = new_rule.contents.to_api_format()['note']
```
Comment (brokensound77) [inline tests/test_transform_fields.py:112]: fix B

**Review by terrancedejesus — APPROVED** [review]
Changes look good!

Tests were fine.
Screenshot 2023-04-24 at 4 30 11 PM

Built a package with `build-release` and checked files that were supposed to have osquery transforms in the investigation guide.

Screenshot 2023-04-24 at 4 34 31 PM

**Pull request merged**